satori/resources/resource_datastore_common.go

```diff
@@ -0,0 +1,306 @@
+						}},
+				},
+			}},
+	}
+}
+
+func createDataStore(d *schema.ResourceData, c *api.Client) (*api.DataStoreOutput, error) {
+	input, err := resourceToDataStore(d)
+
+	if err != nil {
+		return nil, err
+	}
+
+	result, err := c.CreateDataStore(input)
+	if err != nil {
+		return nil, err
+	}
+	d.SetId(result.Id)
+
+	if err := d.Set("id", result.Id); err != nil {
+		return nil, err
+	}
+	return result, err
+}
+
+// convert terraform resource defs into datastore type //
+func resourceToDataStore(d *schema.ResourceData) (*api.DataStore, error) {
+
+	re, err := baselineSecurityPolicyToResource(d.Get("baseline_security_policy").([]interface{}))
+	if err != nil {
+		return nil, err
+	}
+
+	out := api.DataStore{}
+	out.Name = d.Get("name").(string)
+	out.Hostname = d.Get("hostname").(string)
+	out.OriginPort = d.Get(OriginPort).(int)
+	out.CustomIngressPort = d.Get("custom_ingress_port").(int)
+	out.IdentityProviderId = d.Get("identity_provider_id").(string)
+	out.DataAccessControllerId = d.Get("dataaccess_controller_id").(string)
+	out.ProjectIds = convertStringSet(d.Get("project_ids").(*schema.Set))
+	out.BaselineSecurityPolicy = re
+	//} else {
+	//	out.BaselineSecurityPolicy = nil
+	//}
+	out.Type = d.Get("type").(string)
+	return &out, nil
+}
+
+// update datastoreoutput
+func getDataStore(c *api.Client, d *schema.ResourceData) (*api.DataStoreOutput, error) {
+	result, err, statusCode := c.GetDataStore(d.Id())
+	if statusCode == 404 {
+		d.SetId("")
+		return nil, nil
+	}
+	if err != nil {
+		return nil, err
+	}
+
+	d.Set(Id, result.Id)
+	d.Set(Name, result.Name)
+	d.Set(Hostname, result.Hostname)
+	d.Set(ParentId, result.ParentId)
+	d.Set(Type, result.Type)
+	d.Set(OriginPort, result.OriginPort)
+	d.Set(CustomIngressPort, result.CustomIngressPort)
+	d.Set(IdentityProviderId, result.IdentityProviderId)
+	d.Set(DataAccessControllerId, result.DataAccessControllerId)
+	d.Set(ProjectIds, result.ProjectIds)
+
+	tfMap, err := getBaseLinePolicyOutput(result, err)
+	if err != nil {
+		return nil, err
+	}
+	d.Set(BaselineSecurityPolicy, []map[string]interface{}{tfMap})
+	return result, err
+}
+
+func getBaseLinePolicyOutput(result *api.DataStoreOutput, err error) (map[string]interface{}, error) {
+	var inInterface map[string]interface{}
+	inrec, _ := json.Marshal(result.BaselineSecurityPolicy)
+	err = json.Unmarshal(inrec, &inInterface)
+	if err != nil {
+		return nil, err
+	}
+	tfMap := deepCopyMap(inInterface, false)
+	return tfMap, nil
+}
+
+func extractValueFromInterface(in []interface{}) map[string]interface{} {
+	if len(in) > 0 {
+		return in[0].(map[string]interface{})
+	} else {
+		return nil
+	}
+}
+
+func baselineSecurityPolicyToResource(in []interface{}) (*api.BaselineSecurityPolicy, error) {
+	var bls api.BaselineSecurityPolicy
+	lesa := extractValueFromInterface(in)
+	if lesa == nil {
+		return nil, errors.New("no datastore correct")
+	}
+	tfMap := deepCopyMap(lesa, true)
+	jk, _ := json.Marshal(tfMap)
+	err := json.Unmarshal(jk, &bls)
```

Comment:
It looks like kernel reversed engineering code :) can you please give meaningful variable names?